test/integration/network/services_ipfamily_test.go

```diff
@@ -0,0 +1,334 @@
+				if err != nil {
+					t.Errorf("unexpected error getting the services: %v", err)
+				}
+				for _, s := range svcs.Items {
+					// single family Spec.IPFamily must be nil
+					if !tc.dualStack || s.Spec.Type == v1.ServiceTypeExternalName {
+						if s.Spec.IPFamily != nil {
+							t.Errorf("error service %s expected service IP family nil: %v ", s.Name, *s.Spec.IPFamily)
+						}
+					} else if err = validateServiceAndClusterIPFamily(&s, ipFamily); err != nil {
+						t.Errorf("error service %s expected service IP family nil: %v, err: %v ", s.Name, *s.Spec.IPFamily, err)
+					}
+				}
+
+				// Create a service with the non-default IPfamily, it must fail on non-dualstack clusters
+				svc.Name = "svc-ipfamily-other"
+				svc.Spec.IPFamily = &otherIPFamily
+				_, err = client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), svc, metav1.CreateOptions{})
+				if !tc.dualStack && err == nil {
+					t.Errorf("unexpected success creating service: %v", svc)
+				}
+				if tc.dualStack && err != nil {
+					t.Errorf("unexpected error text: %v", err)
+				}
+			})
+		}
+	}
+}
+
+// TestDualStackSkew spawns one API server with DualStack enable and other with DualStack disable
+// and test te behavior of IPFamily with skewed clients
+func TestDualStackSkew(t *testing.T) {
+
+	var testcases = []struct {
+		name            string
+		serviceCIDR     string
+		defaultIPFamily v1.IPFamily
+		dualStack       bool
+		expectErr       string
+	}{
+		{
+			name:            "DualStack enabled and IPv4only",
+			serviceCIDR:     "10.0.0.0/16",
+			defaultIPFamily: v1.IPv4Protocol,
+			dualStack:       true,
+		},
+		{
+			name:            "DualStack enabled and IPv6only",
+			serviceCIDR:     "2001:db8:1::/48",
+			defaultIPFamily: v1.IPv6Protocol,
+			dualStack:       true,
+		},
+		{
+			name:            "DualStack enabled and IPv4 IPv6",
+			serviceCIDR:     "10.0.0.0/16,2001:db8:1::/48",
+			defaultIPFamily: v1.IPv4Protocol,
+			dualStack:       true,
+		},
+		{
+			name:            "DualStack enabled and IPv6 IPv4",
+			serviceCIDR:     "2001:db8:1::/48,10.0.0.0/16",
+			defaultIPFamily: v1.IPv6Protocol,
+			dualStack:       true,
+		},
+	}
+
+	for _, tc := range testcases {
+		tc := tc
+		for _, svc := range serviceTypes {
+			t.Run(fmt.Sprintf("%s/%v", tc.name, svc.Spec.Type), func(t *testing.T) {
+				ipFamily := tc.defaultIPFamily
+				etcd := framework.SharedEtcd()
+
+				instanceOptions := &kubeapiservertesting.TestServerInstanceOptions{
+					DisableStorageCleanup: true,
+				}
+
+				// cleanup the registry storage
+				defer registry.CleanupStorage()
+
+				// start api server without dual stack enabled
+				server := kubeapiservertesting.StartTestServerOrDie(t, instanceOptions, []string{
+					"--endpoint-reconciler-type", "master-count",
+					"--advertise-address", "10.0.1.1",
+					"--apiserver-count", "1",
+					"--feature-gates", "IPv6DualStack=false",
```

Comment:
the servers use a global feature registry... you can't run two servers with different settings for the feature gate in the same process

Reply:
aaargh, :face_with_head_bandage: 

so ... it doesn't seem feasible from the integration framework to do this test ...
:thinking:  how can I simulate a client that's unaware of the ipFamily?